src/msg/async/Event.cc

```diff
@@ -428,13 +428,19 @@ int EventCenter::process_events(unsigned timeout_microseconds,  ceph::timespan *
   if (external_num_events.load()) {
     external_lock.lock();
     deque<EventCallbackRef> cur_process;
+    EventCallbackRef prev_call_back = NULL;
     cur_process.swap(external_events);
     external_num_events.store(0);
     external_lock.unlock();
     numevents += cur_process.size();
     while (!cur_process.empty()) {
       EventCallbackRef e = cur_process.front();
       ldout(cct, 30) << __func__ << " do " << e << dendl;
+      if (prev_call_back == e){
```

Comment:
can we do the dedup in `EventCenter::dispatch_event_external()`? assuming that the worker thread will always be able to pickup the consecutive same events pending in the queue in batch.

Reply:
This situation should be in a scenario where a worker thread has more than one connection and there is only one read-write event at the same time, especially under the pressure of large-scale cluster.

Even with this extreme situation, it can be evaded by configuration switches.

Reply:
@shangfufei sorry, i don't understand your reply. what i suggest is to avoid appending new callbacks to `external_events`, if it already has the same callback at its end. as i think that's the case addressed by your fix.

Reply:
void EventCenter::dispatch_event_external(EventCallbackRef e)
{
  external_lock.lock();
  if(NULL == external_events.find(e))
  {
    external_lock.unlock();
    return;
  }
  external_events.push_back(e);
  bool wake = !external_num_events.load();
  uint64_t num = ++external_num_events;
  external_lock.unlock();
  if (!in_thread() && wake)
    wakeup();

  ldout(cct, 20) << __func__ << " " << e << " pending " << num << dendl;
}
At first I changed this way. but i think he search efficiency of std:: deque is relatively low, which may affect the efficiency of the program.

Reply:
This is  an example of pseudo code.

Reply:
i don't think your pseudo code is equivalent to what you have in this change, instead i'd suggest something like:
```c++
void EventCenter::dispatch_event_external(EventCallbackRef e)
{
  std::lock_guard lock{external_lock};
  bool was_empty = !external_num_events.load();
  if (!was_empty && *external_events.rbegin() == e) {
    return;
  }
  external_events.push_back(e);
  uint64_t num = ++external_num_events;
  if (was_empty && !in_thread()) {
    wakeup();
  }
  ldout(cct, 30) << __func__ << " " << e << " pending " << num << dendl;
}
```

please note, in your change you are always comparing the n and n+1 callback. instead doing the global dedup.

but after a second thought, i wonder, if the thread which

Reply:
I see that the code you modify has the following questions:

1. the scope of lock external_lock will increase. I am not sure whether it will cause other problems.

2. The way the code changes its way of thinking is really improving with my previous changes. The following steps can be omitted: do not put duplicates in the queue, then take them out and release them.

Reply:
void EventCenter::dispatch_event_external(EventCallbackRef e)
{
  external_lock.lock();
  bool was_empty = !external_num_events.load();
  if (!was_empty && *external_events.rbegin() == e) {
  	external_lock.unlock();  
    return;
  }
  external_events.push_back(e);
  uint64_t num = ++external_num_events;
  external_lock.unlock();
  if (was_empty && !in_thread()) {
    wakeup();
  }
  ldout(cct, 30) << __func__ << " " << e << " pending " << num << dendl;
}

Will this be more insurance?

Reply:
> The way the code changes its way of thinking is really improving with my previous changes. The following steps can be omitted: do not put duplicates in the queue, then take them out and release them.

could you rephrase this? and why is this a "problem"?

> the scope of lock external_lock will increase. I am not sure whether it will cause other problems.

if we want to limit the scope of lock, probably we can:

```c++
void EventCenter::dispatch_event_external(EventCallbackRef e)
{
  uint64_t num = 0;
  {
    std::lock_guard lock{external_lock};
    if (external_num_events > 0 && *external_events.rbegin() == e) {
      return;
    }
    external_events.push_back(e);
    num = ++external_num_events;
  }
  if (num == 1 && !in_thread()) {
    wakeup();
  }
  ldout(cct, 30) << __func__ << " " << e << " pending " << num << dendl;
}
```